### PR TITLE
fix(store): improved dispatch typings

### PR DESCRIPTION
**Pull request by splincode**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngxs/store/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the new behavior?

Improved dispatch typings

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

File: packages/store/src/decorators/action.ts [hunk 1]
```diff
@@ -1,29 +1,24 @@
...
+  ActionHandlerMetaData,
+  ActionClass
+} from '../actions/symbols';
 
 /**
  * Decorates a method with a action information.
  */
-export function Action(actions: ActionType | ActionType[], options?: ActionOptions) {
-  return function(target: any, name: string): void {
+export function Action(actions: ActionType | ActionType[], options: ActionOptions = {}) {
```
[the remark this answers is not in the document]
Comment (markwhitfeld) [reply packages/store/src/decorators/action.ts]: I also think that the action parameter type should only support `ActionClass` and `ActionLiteral`.

File: packages/store/src/store.ts [hunk 2]
```diff
@@ -29,7 +30,7 @@ export class Store {
   /**
    * Dispatches event(s).
    */
-  dispatch(event: any | any[]): Observable<any> {
+  dispatch(event: ActionType | ActionType[]): Observable<any> {
```
[the remark this answers is not in the document]
Comment (markwhitfeld) [reply packages/store/src/store.ts]: I think that the action passed in should only support `ActionLiteral` and `ActionStatic`.

File: packages/store/src/symbols.ts [hunk 3]
```diff
@@ -84,7 +85,7 @@ export interface StateContext<T> {
   /**
    * Dispatch a new action and return the dispatched observable.
    */
-  dispatch(actions: any | any[]): Observable<void>;
+  dispatch(actions: ActionType | ActionType[]): Observable<void>;
```
[the remark this answers is not in the document]
Comment (splincode) [reply packages/store/src/symbols.ts]: What do you mean?

**Comment by splincode** [comment]
@markwhitfeld If you know how to fix it on what you are proposing, go to this branch and try. I did everything I could!

**Comment by markwhitfeld** [comment]
The current implementation does not prevent any invalid typings!
```TS
  it('should prevent invalid types passed through', async(() => {
    @State<number>({ name: 'counter', defaults: 0 })
    class MyState {
      @Action(new Increment()) // This should not compile!
      increment({ getState, setState }: StateContext<number>) {
        setState(getState() + 1);
      }
    }

    TestBed.configureTestingModule({
      imports: [NgxsModule.forRoot([MyState])]
    });

    const store: Store = TestBed.get(Store);

    store.dispatch(Increment); // This should not compile!
    store.dispatch({ foo: 123 }); // This should not compile!
  }));
```
image

**Comment by markwhitfeld** [comment]
We can acheive what we want for the `@Action` decorator if we have `ActionType` defined as follows:
```TS
export type ActionType<T = any, U = any> = ActionClass<T, U> | ActionLiteral;
```
Result:
image

**Comment by markwhitfeld** [comment]
The definition (excluding `ActionStatic` is what we had before this PR, so this PR has inadvertently introduced a regression on the type strictness of the `@Action` decorator.

I can't figure out how to get the correct type narrowing on the dispatch method.
The following code can be used as a template of what we need to get working:
```TS
  it('should prevent invalid types passed through', async(() => {
    class Increment {
      static type = 'INCREMENT';
    }
    @State<number>({ name: 'counter', defaults: 0 })
    class MyState {
      @Action(Increment) increment1() {} // This should compile!
      @Action({ type: 'INCREMENT' }) increment2() {} // This should compile!
      @Action(new Increment()) increment3() {} // This should NOT compile!
      @Action({ foo: 123 }) increment4() {} // This should NOT compile!
    }

    TestBed.configureTestingModule({
      imports: [NgxsModule.forRoot([MyState])]
    });

    const store: Store = TestBed.get(Store);

    store.dispatch(new Increment()); // This should compile!
    store.dispatch({ type: 'INCREMENT' }); // This should compile!
    store.dispatch(Increment); // This should NOT compile!
    store.dispatch({ foo: 123 }); // This should NOT compile!
  }));
```

**Comment by markwhitfeld** [comment]
We need to get #925 in place before we can reasonably look at this type of PR. Just too much risk.

File: packages/store/src/modules/ngxs-root.module.ts [hunk 4]
```diff
@@ -16,14 +16,14 @@ import { InitState } from '../actions/actions';
...
     @Optional()
     @Inject(ROOT_STATE_TOKEN)
-    states: StateClass[] = [],
-    lifecycleStateManager: LifecycleStateManager
+    public states: StateClass[] = [],
+    public factory: StateFactory,
+    public internalStateOperations: InternalStateOperations,
+    public store: Store,
+    public select: SelectFactory,
+    public lifecycleStateManager: LifecycleStateManager
```
Comment (markwhitfeld) [inline packages/store/src/modules/ngxs-root.module.ts]: These should not be public

File: packages/store/src/modules/ngxs-feature.module.ts [hunk 5]
```diff
@@ -15,13 +15,13 @@ import { UpdateState } from '../actions/actions';
...
-    factory: StateFactory,
     @Optional()
     @Inject(FEATURE_STATE_TOKEN)
-    states: StateClass[][] = [],
-    lifecycleStateManager: LifecycleStateManager
+    public states: StateClass[][] = [],
+    public store: Store,
+    public internalStateOperations: InternalStateOperations,
+    public factory: StateFactory,
+    public lifecycleStateManager: LifecycleStateManager
```
Comment (markwhitfeld) [inline packages/store/src/modules/ngxs-feature.module.ts]: These should not be public

File: packages/store/src/execution/dispatch-outside-zone-ngxs-execution-strategy.ts [hunk 6]
```diff
@@ -39,14 +31,4 @@ export class DispatchOutsideZoneNgxsExecutionStrategy implements NgxsExecutionSt
...
-
-  private verifyZoneIsNotNooped(ngZone: NgZone): void {
-    /* - Removed because unsafe for Angular 5 - investigate
-    if (ngZone instanceof NoopNgZone) {
-      console.warn(
-        'Your application was bootstrapped with nooped zone and your execution strategy requires an ngZone'
-      );
-    }
-    */
-  }
```
Comment (markwhitfeld) [inline packages/store/src/execution/dispatch-outside-zone-ngxs-execution-strategy.ts]: Please leave this in if it requires investigation.

**Comment by markwhitfeld** [comment]
@splincode The work you have done here to add typing tests is great, but would it be possible to add those in a seperate PR, and then once that is merged we would be able to see how the typings changed in this PR.

**Comment by splincode** [comment]
outdated

**Pull request closed without merging**